### PR TITLE
Межуев Максим Задача 3 Вариант 29 "Выделение ребер на изображении с использованием оператора Собеля"

### PR DESCRIPTION
**Pull request by Domovenokkk**
В рамках данной задачи требуется разработать и протестировать алгоритм выделения границ методом Собеля с использованием MPI  для параллельных вычислений. Реализация включает работу с входными данными (изображениями), их обработку, распределение задач между процессами MPI и сбор результатов.

### CODE REVIEW

**Review by Kvoks — CHANGES_REQUESTED** [review]
code coverage is too low

**Review by Kvoks — APPROVED** [review]

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  }
+  return true;
+}
+
+bool SobelEdgeDetectionMPI::pre_processing(TaskData* task_data) {
+  if (task_data == nullptr) {
+    return false;
+  }
+
+  if (!validation()) {
```
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:38]: this is not needed here

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  taskData = task_data;
+
+  gradient_x.resize(taskData->width * taskData->height);
+  gradient_y.resize(taskData->width * taskData->height);
+
+  return true;
+}
+
+bool SobelEdgeDetectionMPI::run() {
+  if (taskData == nullptr) {
```
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:51]: its validation

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  }
+  if (taskData->inputs_count.empty() || taskData->outputs_count.empty() ||
+      taskData->inputs_count[0] != taskData->outputs_count[0]) {
+    return false;
+  }
+  return true;
+}
+
+bool SobelEdgeDetectionMPI::pre_processing(TaskData* task_data) {
+  if (task_data == nullptr) {
```
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:34]: its validation

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  size_t height = taskData->height;
+
+  if (width == 0 || height == 0) {
+    return false;
+  }
+
+  uint8_t* input_image = taskData->inputs[0];
+  uint8_t* output_image = taskData->outputs[0];
+
+  if (input_image == nullptr || output_image == nullptr) {
```
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:67]: its validation

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+  if (taskData == nullptr) {
+    return false;
+  }
+
+  int rank = world.rank();
+  int size = world.size();
+  size_t width = taskData->width;
+  size_t height = taskData->height;
+
+  if (width == 0 || height == 0) {
```
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:60]: its validation

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    return false;
+  }
+
+  uint8_t* input_image = taskData->inputs[0];
+  uint8_t* output_image = taskData->outputs[0];
```
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:65]: its preproccesing

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+      output_image[y * width + x] = std::min(magnitude, 255);
+    }
+  }
+
+  world.barrier();
+  return true;
+}
+
+bool SobelEdgeDetectionMPI::post_processing() {
+  if (taskData == nullptr || taskData->outputs[0] == nullptr) {
```
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:103]: its validation

**Review by decl1 — CHANGES_REQUESTED** [review]

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+#include <cmath>
+#include <iostream>
+#include <vector>
+
+namespace mezhuev_m_sobel_edge_detection {
+
+bool SobelEdgeDetectionMPI::setTaskData(TaskData* task_data) {
+  taskData = task_data;
+  return true;
+}
```
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:14]: I think you shouldn't do this. The class must be inherited

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+      taskData->inputs_count[0] != taskData->outputs_count[0]) {
+    return false;
+  }
+  return true;
+}
+
+bool SobelEdgeDetectionMPI::pre_processing(TaskData* task_data) {
+  if (task_data == nullptr) {
+    return false;
+  }
```
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:36]: this is validation

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+  if (!validation()) {
+    return false;
+  }
```
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:40]: u dont need do this check

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+  gradient_x.resize(taskData->width * taskData->height);
+  gradient_y.resize(taskData->width * taskData->height);
+
+  return true;
+}
+
+bool SobelEdgeDetectionMPI::run() {
+  if (taskData == nullptr) {
+    return false;
+  }
```
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:53]: this is validation

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+  int rank = world.rank();
+  int size = world.size();
+  size_t width = taskData->width;
+  size_t height = taskData->height;
+
+  if (width == 0 || height == 0) {
+    return false;
+  }
```
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:62]: this is validation

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+  uint8_t* input_image = taskData->inputs[0];
+  uint8_t* output_image = taskData->outputs[0];
+
+  if (input_image == nullptr || output_image == nullptr) {
+    return false;
+  }
```
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:69]: this is validation

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp (hunk 6 above)
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:65]: its preproccesing

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+          gy += sobel_y[ky + 1][kx + 1] * pixel_value;
+        }
+      }
+
+      int magnitude = static_cast<int>(std::sqrt(gx * gx + gy * gy));
+      output_image[y * width + x] = std::min(magnitude, 255);
+    }
+  }
+
+  world.barrier();
```
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:98]: I could be wrong. it is not necessary to put a barrier before return

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp (hunk 7 above)
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:103]: its validation

File: tasks/mpi/mezhuev_m_sobel_edge_detection/include/ops_mpi.hpp [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+  bool post_processing();
+
+  bool setTaskData(TaskData* task_data);
+
+ private:
+  boost::mpi::communicator& world;
+  TaskData* taskData = nullptr;
+  std::vector<int> gradient_x;
+  std::vector<int> gradient_y;
+};
```
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/include/ops_mpi.hpp:42]: this class must be inherited : public ppc::core::Task

File: tasks/mpi/mezhuev_m_sobel_edge_detection/include/ops_mpi.hpp [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+namespace mezhuev_m_sobel_edge_detection {
+
+struct TaskData {
+  size_t width;
+  size_t height;
+  std::vector<uint8_t*> inputs;
+  std::vector<uint8_t*> outputs;
+  std::vector<size_t> inputs_count;
+  std::vector<size_t> outputs_count;
+};
```
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/include/ops_mpi.hpp:24]: it's in public ppc::core::Task

File: tasks/mpi/mezhuev_m_sobel_edge_detection/include/ops_mpi.hpp (hunk 15 above)
Comment (decl1) [reply tasks/mpi/mezhuev_m_sobel_edge_detection/include/ops_mpi.hpp:42]: and you must add seq task in ops_mpi.hpp. You need compare them in func tests.

**Review by decl1 — CHANGES_REQUESTED** [review]

File: tasks/mpi/mezhuev_m_sobel_edge_detection/include/ops_mpi.hpp (hunk 15 above)
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/include/ops_mpi.hpp:42]: you tasks classes must be from public ppc::core::Task

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp (hunk 8 above)
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:14]: must be from public ppc::core::Task

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+
+namespace mezhuev_m_sobel_edge_detection {
+
+bool SobelEdgeDetectionMPI::setTaskData(TaskData* task_data) {
+  taskData = task_data;
+  return true;
+}
+
+bool SobelEdgeDetectionMPI::validation() {
+  if (taskData == nullptr || taskData->inputs.empty() || taskData->outputs.empty()) {
```
Comment (decl1) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:17]: in begin of every task function must be internal_order_test();

**Review by decl1 — CHANGES_REQUESTED** [review]

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp (hunk 17 above)
Comment (decl1) [reply tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:17]:
```suggestion
bool SobelEdgeDetectionMPI::validation() {
  internal_order_test();
  
  // your function code here
  
  return ...;
}
```

**Review by decl1 — APPROVED** [review]

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+  if (taskData->inputs.size() != 1 || taskData->outputs.size() != 1) {
+    return false;
+  }
+  if (taskData->inputs[0] == nullptr || taskData->outputs[0] == nullptr) {
+    return false;
+  }
+  if (taskData->inputs_count.empty() || taskData->outputs_count.empty()) {
+    return false;
+  }
+  if (taskData->inputs_count[0] != taskData->outputs_count[0]) {
```
Comment (allnes) [inline tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:25]: where check sizes of inputs_count, outputs_count?

File: tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp (hunk 6 above)
Comment (allnes) [reply tasks/mpi/mezhuev_m_sobel_edge_detection/src/ops_mpi.cpp:52]: @Domovenokkk you haven't corrected this comment

**Comment by allnes** [comment]
@decl1 you approved this PR, but some comments wasn't corrected, @Domovenokkk please add 1 addtional reviewer

**Comment by aobolensk** [comment]
Please increase test coverage

**Pull request closed without merging**